File: android/src/main/kotlin/com/hypertrack/sdk/flutter/common/Result.kt
package com.hypertrack.sdk.flutter.common

sealed class Result<S> {
    fun <T> flatMap(function: (S) -> Result<T>): Result<T> {
        return when(this) {
            is Success -> {
                function.invoke(this.success)
            }
            is Failure -> {
                Failure<T>(this.failure)
            }
        }
    }
}
data class Success<S>(val success: S): Result<S>()
data class Failure<S>(val failure: Exception): Result<S>()

File: lib/src/serialization/result.dart
import '../../data_types/result.dart';
import 'common.dart';

Result<Map<Object?, Object?>, Map<Object?, Object?>> deserializeResult(
    Map<Object?, Object?> response) {
  try {
    Map<String, dynamic> data = response.cast<String, dynamic>();
    switch (data[keyType]) {
      case _typeSuccess:
        return Result.success(data[keyValue]);
      case _typeFailure:
        return Result.error(data[keyValue]);
      default:
        throw Exception("Invalid location response: ${response}");
    }
  } catch (e) {
    throw Exception("Invalid location response: ${response} $e");
  }
}

const _typeSuccess = "success";
const _typeFailure = "failure";
